plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/Utils.hs

```diff
@@ -311,38 +315,118 @@ costIsAcceptable = \case
+                decSize >> go body
+            Constr _ _ _ es ->
+                -- `decSize` twice to account for the `Constr` node and the index.
+                decSize >> decSize >> allM go es
+            Case _ _ arg cs -> decSize >> allM go (arg : cs)
+            IWrap _ _ _ body ->
+                -- No `decSize` here, because `IWrap` nodes disappear in UPLC.
+                go body
+            Unwrap _ body ->
+                -- No `decSize` here, because `Unwrap` nodes disappear in UPLC.
+                go body
+            -- Constants can be big! We could check the size here and inline if they're
+            -- small, but probably not worth it
+            Constant{} -> pure False
+            Apply _ fun arg ->
+                -- `decSize` to account for the `Apply` node.
+                decSize >> (go fun &&^ go arg)
+            TyInst _ body _ty ->
+                -- `decSize` to account for the `TyInst` node (which becomes `Force` in UPLC).
+                decSize >> go body
+            Let{} -> pure False
+
+    decSize = modify' (over stRemainingSize (\x -> x - 1))
 
 -- | Is this an utterly trivial type which might as well be inlined?
 trivialType :: Type tyname uni ann -> Bool
 trivialType = \case
     TyBuiltin{} -> True
     TyVar{}     -> True
     _           -> False
+
+{- Note [Size criteria for inlining]
+Inlining replaces LHS with RHS, for some `LHS = RHS`. The LHS is usually a variable and the
+RHS is usually the definition of the variable. But when we are considering inlining
+fully-saturated applications, LHS can have parameters.
+
+For example, given `x = \a b c d -> f y`, normally LHS is `x`. But when we are considering
+inlining a call site of `x` where `x` is fully applied, LHS becomes `x a b c d`, and
+RHS becomes `f y`.
+
+In order to avoid increasing program sizes (which is much more important for Plutus Core than
+it is for general-purpose programming languages), we should inline only if size(RHS) is
+no bigger than size(LHS).
+
+size(LHS) depends on the number of term parameters on the LHS. For `x = f y`, size(LHS) = 1.
+Each additional parameter increases size(LHS) by 2 (one for the `Apply` node, and one for
+the parameter itself). For `x a b c d = f y`, size(LHS) = 9.
+
+We ignore type parameters since types eventually get erased.
+
+To measure size(RHS), we have to be mindful of multiple occurrences of parameters.
+For example, in `x a b c = b a a`, we cannot consider size(RHS) to be no bigger than size(LHS)
+just because it is not bigger in the binding. Otherwise we will turn `x <large> b c` into
+`let a = <large> in b a a`, which can be worse than not inlining.
+
+We therefore only consider size(RHS) to be acceptable for inlining, if no parameter in the LHS
+occurs more than once in the RHS, or, to put it another way, if every parameter on the LHS can
+be unconditionally inlined after beta-reduction. This is not the optimal strategy but
+it balances simplicity and usefulness.
```

Comment:
Okay, so I modify my suggestion from before. We can combine two, simple checks:
- Compare `size (original term)` and `size (body)`. The difference must be below some threshold (later relaxed/configurable)
- Check that the number of occurrences of the bound variables is 1. We already have a usages map for checking this, so no need to do something clever again. This is replicating part of our other inlining heuristics, so it _should_ use the same code.

Reply:
> Compare size (original term) and size (body).

Hmm, that's right. Somehow I overlooked this option.